File: ws/first_ws/src/service_pkg/scripts/server_custom_srv.py
#! /usr/bin/env python
# coding:utf8

import rospy
from demo_srvs.srv import NumOption, NumOptionRequest, NumOptionResponse


def client_callback(request):
    if not isinstance(request, NumOptionRequest):
        return

    response = NumOptionResponse()
    option = request.option

    if option == "+":
        response.result = request.a + request.b
    elif option == "-":
        response.result = request.a - request.b
    elif option == "*":
        response.result = request.a * request.b
    elif option == "/":
        response.result = request.a / request.b
    elif option == "%":
        response.result = request.a % request.b
    else:
        return None

    return response


if __name__ == '__main__':
    nodeName = "python_service_node"
    rospy.init_node(nodeName)

    serviceName = "/python/service"

    # 与c++ api 不一致，
    server = rospy.Service(serviceName, NumOption, client_callback)

    rospy.spin()

File: ros/ws01/first_ws/src/service_pkg/src/server_custom_srv.cpp
//
// Created by zcb on 2021/4/9.
//


#include <ros/ros.h>
#include <iostream>
#include <demo_srvs/NumOption.h>

using namespace std;

bool client_callback(demo_srvs::NumOptionRequest &request, demo_srvs::NumOptionResponse &response) {
    string option = request.option;
    if (option == "+") {
        response.result = request.a + request.b;
    } else if (option == "-") {
        response.result = request.a - request.b;
    } else if (option == "*") {
        response.result = request.a * request.b;
    } else if (option == "/") {
        response.result = request.a / request.b;
    } else if (option == "%") {
        response.result = fmod(request.a, request.b);
    } else {
        return false;
    }
    return true;
}


int main(int argc, char **argv) {

    string nodeName = "cpp_server_node";
    ros::init(argc, argv, nodeName);
    ros::NodeHandle node;

    string serviceName = "/cpp/service";
    const ros::ServiceServer &server = node.advertiseService(serviceName, client_callback);

    ros::spin();

    return 0;
}








